File: Python/1074-reversing_linked_list.py


class Node():
    def __init__(self, addr, data, nxt):
        self.addr = addr
        self.data = data
        self.nxt = nxt

    def __str__(self):
        return f'{self.addr} {self.data} {self.nxt}'


def reversing_linked_list(linked_list, k):
    n_node = len(linked_list)
    r_linked_list = []
    last_node = Node(None, None, None)
    for i in range(0, n_node, k):
        if i + k <= n_node:
            for node in reversed(linked_list[i:i+k]):
                last_node.nxt = node.addr
                r_linked_list.append(node)
                last_node = node
        else:
            last_node.nxt = linked_list[i].addr
            r_linked_list.extend(linked_list[i:i+k])
    r_linked_list[-1].nxt = '-1'  # the end of the list
    return r_linked_list


if __name__ == '__main__':
    head, n_node, k = input().split()
    n_node, k = int(n_node), int(k)
    nodes = {}
    for _ in range(n_node):
        addr, data, nxt = input().split()
        nodes[addr] = Node(addr, data, nxt)
    linked_list = []
    nxt = head
    while nxt != '-1':
        node = nodes[nxt]
        linked_list.append(node)
        nxt = node.nxt
    r_linked_list = reversing_linked_list(linked_list, k)
    for node in r_linked_list:
        print(node)

File: C++/1074-reversing_linked_list.cc
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>


struct Node {
    std::string addr, data, next;
};


std::vector<Node>
reversing_linked_list(const std::vector<Node> &linked_list, int k)
{
    int n_node = linked_list.size();
    std::vector<Node> r_linked_list;
    int count = 0;
    for (int i=0; i<n_node; i+=k) {
        if (i + k <= n_node) {
            for (int r_i=i+k-1; r_i>=i; --r_i) {
                Node node = linked_list[r_i];
                if (count > 0) {
                    r_linked_list[count-1].next = node.addr;
                }
                r_linked_list.push_back(node);
                ++count;
            }
        } else {
            if (count > 0) {
                r_linked_list[count-1].next = linked_list[i].addr;
            }
            for (int r_i=i; r_i<n_node; ++r_i) {
                r_linked_list.push_back(linked_list[r_i]);
            }
        }
    }
    r_linked_list[n_node-1].next = "-1";
    return r_linked_list;
}


int main() {

    std::ios::sync_with_stdio(false);

    std::string head;
    int n_node, k;
    std::cin >> head >> n_node >> k;

    std::unordered_map<std::string, Node> nodes;
    while (n_node--) {
        std::string addr, data, next;
        std::cin >> addr >> data >> next;
        nodes[addr] = {addr, data, next};
    }

    std::vector<Node> linked_list;
    std::string next = head;
    while (next != "-1") {
        Node node = nodes[next];
        linked_list.push_back(node);
        next = node.next;
    }

    auto r_linked_list = reversing_linked_list(linked_list, k);
    for (Node &node : r_linked_list) {
        std::cout << node.addr << ' ' << node.data << ' ' << node.next << '\n';
    }

    return 0;
}
